include/clink/linalg/sparse_vector.h

```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2021 The Clink Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef SPARSE_VECTOR_H_
+#define SPARSE_VECTOR_H_
+
+namespace clink {
+
+// A sparse vector of double values.
+typedef struct SparseVector {
+  SparseVector(const int, const int *, const int, const double *, const int);
+  ~SparseVector();
+
+  int n;
+  int *indices;
```

Comment:
It is not very safe and readable to represent a dynamic array of int in this way. And this approach has worse performance than defining the variables inline.

How about we use `llvm::SmallVector<int, 2>` here? See https://stackoverflow.com/questions/42122054/what-is-the-difference-between-stdvector-and-llvmsmallvector-which-one-to-u for related discussion.

Same for `values`.

Reply:
I agree that we should use existing C++ array-like classes like `llvm::SmallVector<int, 2>`. In pure C++'s case this could help improve effciency.
However, As described in JNA's [Javadoc](https://java-native-access.github.io/jna/4.2.0/overview-summary.html#marshalling), JNA does not support custom C++ classes. The closest approach is to make `SparseVector` a class and pass its `Pointer` to Java. Then the Java program can extract information about the SparseVector using its get() methods. I'll make the change in this way.